From c8415f6f2271008aef5056689950236df627d9b1 Mon Sep 17 00:00:00 2001
From: lijuang <lijuang@codeaurora.org>
Date: Tue, 10 Oct 2017 21:33:20 +0800
Subject: QcomModulePkg: Fix the array index out of bounds error for
 UsbDescriotor

The index of the array StrSerialDescriptor maybe is out of bounds

Change-Id: I7a64339ff761d44fcba2407aad30f814e480cc8b
---
 QcomModulePkg/Library/FastbootLib/UsbDescriptors.c | 22 ++++++++++++++++++++--
 1 file changed, 20 insertions(+), 2 deletions(-)

diff --git a/QcomModulePkg/Library/FastbootLib/UsbDescriptors.c b/QcomModulePkg/Library/FastbootLib/UsbDescriptors.c
index e3f4cf9..678ace9 100644
--- a/QcomModulePkg/Library/FastbootLib/UsbDescriptors.c
+++ b/QcomModulePkg/Library/FastbootLib/UsbDescriptors.c
@@ -271,8 +271,8 @@ BuildDefaultDescriptors(
   OUT VOID                  **SSDescriptors)
 {
   UINT8                        Index = 0;
-  UINT8                        NumCfg = 0; 
-  CHAR8                        Str_UUID[64];
+  UINT8                        NumCfg = 0;
+  CHAR8                        Str_UUID[UUID_STR_LEN];
   UINT32                       i;
   EFI_STATUS                   Status;
 
@@ -283,6 +283,24 @@ BuildDefaultDescriptors(
     return;
   }
 
+  /* Full UUID descriptor should be length 74, now it only
+   * works up to 62.
+   * The array members of StrSerialDescriptor is:
+   * sizeof(StrSerialDescriptor), USB_DESC_TYPE_STRING,
+   * '9', 0,
+   * 'a', 0,
+   * '1', 0,
+   * '8', 0,
+   * '9', 0,
+   * '9', 0,
+   * '1', 0,
+   */
+  if (((AsciiStrLen (Str_UUID) - 1) * 2 + 3) > (MAX_DESC_LEN - 1)) {
+     DEBUG ((EFI_D_ERROR,
+         "Error the array index out of bounds\n"));
+     return;
+  }
+
   StrSerialDescriptor[0] = AsciiStrLen(Str_UUID) * 2 + 2;
   StrSerialDescriptor[1] = USB_DESC_TYPE_STRING;
   for (i = 0; i < AsciiStrLen(Str_UUID); i++) {
-- 
cgit v1.1

